File: tests/system_tests/memory_allocation/memory_allocation.rs
use std::env;
use std::fs::File;
use std::io::BufReader;
use std::time::{SystemTime, UNIX_EPOCH, Instant};
use std::collections::HashMap;
use serde::{Deserialize, Serialize};
use serde_json;
use rand::Rng;

#[derive(Debug, Deserialize)]
struct Config {
    parameters: Parameters,
}

#[derive(Debug, Deserialize)]
struct Parameters {
    allocation_sizes: Vec<usize>,
    allocation_patterns: Vec<String>,
    allocation_counts: Vec<usize>,
    data_structures: Vec<String>,
    iterations: usize,
}

#[derive(Debug, Serialize)]
struct Results {
    start_time: f64,
    test_cases: Vec<TestCase>,
    summary: Summary,
    end_time: f64,
    total_execution_time: f64,
}

#[derive(Debug, Serialize)]
struct TestCase {
    allocation_size: usize,
    allocation_count: usize,
    data_structure: String,
    allocation_pattern: String,
    iterations: Vec<IterationResult>,
    avg_allocation_time: f64,
    avg_deallocation_time: f64,
    avg_memory_efficiency: f64,
}

#[derive(Debug, Serialize)]
struct IterationResult {
    iteration: usize,
    initial_memory: usize,
    allocation: AllocationResult,
    deallocation: DeallocationResult,
}

#[derive(Debug, Serialize)]
struct AllocationResult {
    success: bool,
    time_ms: f64,
    memory_used: usize,
    peak_memory: usize,
    memory_efficiency: f64,
    items_allocated: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    error: Option<String>,
}

#[derive(Debug, Serialize)]
struct DeallocationResult {
    success: bool,
    time_ms: f64,
    final_memory: usize,
    memory_freed: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    error: Option<String>,
}

#[derive(Debug, Serialize)]
struct Summary {
    total_tests: usize,
    successful_tests: usize,
    failed_tests: usize,
    avg_allocation_time: f64,
    avg_deallocation_time: f64,
    avg_memory_efficiency: f64,
}

// Simple memory tracking (approximation)
fn get_memory_usage() -> usize {
    // In Rust, we'll use a simple approximation since direct memory monitoring is complex
    // This is a placeholder - in real scenarios you'd use external crates like sysinfo
    0 // Returning 0 as placeholder
}

fn allocate_arrays(size: usize, count: usize) -> Vec<Vec<i32>> {
    let mut rng = rand::thread_rng();
    let mut arrays = Vec::with_capacity(count);
    
    for _ in 0..count {
        let mut array = Vec::with_capacity(size);
        for _ in 0..size {
            array.push(rng.gen_range(0..1000));
        }
        arrays.push(array);
    }
    
    arrays
}

fn allocate_hash_maps(size: usize, count: usize) -> Vec<HashMap<i32, i32>> {
    let mut rng = rand::thread_rng();
    let mut maps = Vec::with_capacity(count);
    
    for _ in 0..count {
        let mut map = HashMap::with_capacity(size);
        for _ in 0..size {
            let key = rng.gen_range(0..size * 2) as i32;
            let value = rng.gen_range(0..1000);
            map.insert(key, value);
        }
        maps.push(map);
    }
    
    maps
}

#[derive(Clone)]
struct ListNode {
    value: i32,
    next: Option<Box<ListNode>>,
}

fn allocate_linked_lists(size: usize, count: usize) -> Vec<Option<Box<ListNode>>> {
    let mut rng = rand::thread_rng();
    let mut lists = Vec::with_capacity(count);
    
    for _ in 0..count {
        let mut head: Option<Box<ListNode>> = None;
        for _ in 0..size {
            let new_node = ListNode {
                value: rng.gen_range(0..1000),
                next: head,
            };
            head = Some(Box::new(new_node));
        }
        lists.push(head);
    }
    
    lists
}

fn run_memory_allocation_benchmark(params: Parameters) -> Results {
    let start_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs_f64();
    let mut test_cases = Vec::new();
    let mut summary = Summary {
        total_tests: 0,
        successful_tests: 0,
        failed_tests: 0,
        avg_allocation_time: 0.0,
        avg_deallocation_time: 0.0,
        avg_memory_efficiency: 0.0,
    };
    
    let mut all_allocation_times = Vec::new();
    let mut all_deallocation_times = Vec::new();
    let mut all_memory_efficiencies = Vec::new();
    
    for &size in &params.allocation_sizes {
        for &count in &params.allocation_counts {
            for structure in &params.data_structures {
                for pattern in &params.allocation_patterns {
                    eprintln!("Testing {} allocation: size={}, count={}, pattern={}...", 
                             structure, size, count, pattern);
                    
                    let mut test_case = TestCase {
                        allocation_size: size,
                        allocation_count: count,
                        data_structure: structure.clone(),
                        allocation_pattern: pattern.clone(),
                        iterations: Vec::new(),
                        avg_allocation_time: 0.0,
                        avg_deallocation_time: 0.0,
                        avg_memory_efficiency: 0.0,
                    };
                    
                    let mut allocation_times = Vec::new();
                    let mut deallocation_times = Vec::new();
                    let mut memory_efficiencies = Vec::new();
                    
                    for i in 0..params.iterations {
                        eprintln!("  Iteration {}/{}...", i + 1, params.iterations);
                        
                        let initial_memory = get_memory_usage();
                        summary.total_tests += 1;
                        
                        let mut iteration_result = IterationResult {
                            iteration: i + 1,
                            initial_memory,
                            allocation: AllocationResult {
                                success: false,
                                time_ms: 0.0,
                                memory_used: 0,
                                peak_memory: 0,
                                memory_efficiency: 0.0,
                                items_allocated: count,
                                error: None,
                            },
                            deallocation: DeallocationResult {
                                success: false,
                                time_ms: 0.0,
                                final_memory: 0,
                                memory_freed: 0,
                                error: None,
                            },
                        };
                        
                        let success = match structure.as_str() {
                            "array" => {
                                let start = Instant::now();
                                let _allocated = allocate_arrays(size, count);
                                let allocation_time = start.elapsed().as_secs_f64() * 1000.0;
                                
                                let peak_memory = get_memory_usage();
                                let memory_used = peak_memory.saturating_sub(initial_memory);
                                let theoretical_size = size * count * std::mem::size_of::<i32>();
                                let memory_efficiency = if memory_used > 0 {
                                    (theoretical_size as f64 / memory_used as f64) * 100.0
                                } else {
                                    100.0
                                };
                                
                                allocation_times.push(allocation_time);
                                all_allocation_times.push(allocation_time);
                                memory_efficiencies.push(memory_efficiency);
                                all_memory_efficiencies.push(memory_efficiency);
                                
                                iteration_result.allocation = AllocationResult {
                                    success: true,
                                    time_ms: allocation_time,
                                    memory_used,
                                    peak_memory,
                                    memory_efficiency,
                                    items_allocated: count,
                                    error: None,
                                };
                                
                                // Deallocation (drop happens automatically)
                                let start = Instant::now();
                                drop(_allocated);
                                let deallocation_time = start.elapsed().as_secs_f64() * 1000.0;
                                let final_memory = get_memory_usage();
                                
                                deallocation_times.push(deallocation_time);
                                all_deallocation_times.push(deallocation_time);
                                
                                iteration_result.deallocation = DeallocationResult {
                                    success: true,
                                    time_ms: deallocation_time,
                                    final_memory,
                                    memory_freed: peak_memory.saturating_sub(final_memory),
                                    error: None,
                                };
                                
                                true
                            },
                            "hash_map" => {
                                let start = Instant::now();
                                let _allocated = allocate_hash_maps(size, count);
                                let allocation_time = start.elapsed().as_secs_f64() * 1000.0;
                                
                                let peak_memory = get_memory_usage();
                                let memory_used = peak_memory.saturating_sub(initial_memory);
                                let theoretical_size = size * count * (std::mem::size_of::<i32>() * 2);
                                let memory_efficiency = if memory_used > 0 {
                                    (theoretical_size as f64 / memory_used as f64) * 100.0
                                } else {
                                    100.0
                                };
                                
                                allocation_times.push(allocation_time);
                                all_allocation_times.push(allocation_time);
                                memory_efficiencies.push(memory_efficiency);
                                all_memory_efficiencies.push(memory_efficiency);
                                
                                iteration_result.allocation = AllocationResult {
                                    success: true,
                                    time_ms: allocation_time,
                                    memory_used,
                                    peak_memory,
                                    memory_efficiency,
                                    items_allocated: count,
                                    error: None,
                                };
                                
                                let start = Instant::now();
                                drop(_allocated);
                                let deallocation_time = start.elapsed().as_secs_f64() * 1000.0;
                                let final_memory = get_memory_usage();
                                
                                deallocation_times.push(deallocation_time);
                                all_deallocation_times.push(deallocation_time);
                                
                                iteration_result.deallocation = DeallocationResult {
                                    success: true,
                                    time_ms: deallocation_time,
                                    final_memory,
                                    memory_freed: peak_memory.saturating_sub(final_memory),
                                    error: None,
                                };
                                
                                true
                            },
                            "linked_list" => {
                                let start = Instant::now();
                                let _allocated = allocate_linked_lists(size, count);
                                let allocation_time = start.elapsed().as_secs_f64() * 1000.0;
                                
                                let peak_memory = get_memory_usage();
                                let memory_used = peak_memory.saturating_sub(initial_memory);
                                let theoretical_size = size * count * (std::mem::size_of::<i32>() + std::mem::size_of::<usize>());
                                let memory_efficiency = if memory_used > 0 {
                                    (theoretical_size as f64 / memory_used as f64) * 100.0
                                } else {
                                    100.0
                                };
                                
                                allocation_times.push(allocation_time);
                                all_allocation_times.push(allocation_time);
                                memory_efficiencies.push(memory_efficiency);
                                all_memory_efficiencies.push(memory_efficiency);
                                
                                iteration_result.allocation = AllocationResult {
                                    success: true,
                                    time_ms: allocation_time,
                                    memory_used,
                                    peak_memory,
                                    memory_efficiency,
                                    items_allocated: count,
                                    error: None,
                                };
                                
                                let start = Instant::now();
                                drop(_allocated);
                                let deallocation_time = start.elapsed().as_secs_f64() * 1000.0;
                                let final_memory = get_memory_usage();
                                
                                deallocation_times.push(deallocation_time);
                                all_deallocation_times.push(deallocation_time);
                                
                                iteration_result.deallocation = DeallocationResult {
                                    success: true,
                                    time_ms: deallocation_time,
                                    final_memory,
                                    memory_freed: peak_memory.saturating_sub(final_memory),
                                    error: None,
                                };
                                
                                true
                            },
                            _ => {
                                iteration_result.allocation.error = Some(format!("Unknown data structure: {}", structure));
                                false
                            }
                        };
                        
                        if success {
                            summary.successful_tests += 1;
                        } else {
                            summary.failed_tests += 1;
                        }
                        
                        test_case.iterations.push(iteration_result);
                    }
                    
                    // Calculate averages
                    if !allocation_times.is_empty() {
                        test_case.avg_allocation_time = allocation_times.iter().sum::<f64>() / allocation_times.len() as f64;
                    }
                    if !deallocation_times.is_empty() {
                        test_case.avg_deallocation_time = deallocation_times.iter().sum::<f64>() / deallocation_times.len() as f64;
                    }
                    if !memory_efficiencies.is_empty() {
                        test_case.avg_memory_efficiency = memory_efficiencies.iter().sum::<f64>() / memory_efficiencies.len() as f64;
                    }
                    
                    test_cases.push(test_case);
                }
            }
        }
    }
    
    // Calculate overall summary
    if !all_allocation_times.is_empty() {
        summary.avg_allocation_time = all_allocation_times.iter().sum::<f64>() / all_allocation_times.len() as f64;
    }
    if !all_deallocation_times.is_empty() {
        summary.avg_deallocation_time = all_deallocation_times.iter().sum::<f64>() / all_deallocation_times.len() as f64;
    }
    if !all_memory_efficiencies.is_empty() {
        summary.avg_memory_efficiency = all_memory_efficiencies.iter().sum::<f64>() / all_memory_efficiencies.len() as f64;
    }
    
    let end_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs_f64();
    
    Results {
        start_time,
        test_cases,
        summary,
        end_time,
        total_execution_time: end_time - start_time,
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <config_file>", args[0]);
        std::process::exit(1);
    }
    
    let config_file = &args[1];
    let file = File::open(config_file)?;
    let reader = BufReader::new(file);
    let config: Config = serde_json::from_reader(reader)?;
    
    let results = run_memory_allocation_benchmark(config.parameters);
    
    println!("{}", serde_json::to_string_pretty(&results)?);
    
    Ok(())
}

File: tests/system_tests/memory_allocation/memory_allocation.py
#!/usr/bin/env python3
"""
Memory allocation test implementation in Python.
Measures memory allocation, deallocation, and management performance.
"""

import json
import sys
import time
import random
import gc
import psutil
import os
from typing import Dict, List, Any, Optional
import numpy as np


def get_memory_usage() -> int:
    """Get current memory usage in bytes."""
    try:
        process = psutil.Process(os.getpid())
        return process.memory_info().rss
    except:
        return 0


def allocate_array(size: int, count: int) -> List[np.ndarray]:
    """Allocate arrays of specified size using NumPy for performance."""
    arrays = []
    for i in range(count):
        # Use NumPy's optimized random integer generation
        array = np.random.randint(0, 1001, size=size)
        arrays.append(array)
    return arrays


def allocate_linked_list(size: int, count: int) -> List[List[Dict]]:
    """Allocate linked list structures."""
    class ListNode:
        def __init__(self, value: int, next_node=None):
            self.value = value
            self.next = next_node
    
    lists = []
    for i in range(count):
        head = None
        for j in range(size):
            new_node = ListNode(random.randint(0, 1000))
            new_node.next = head
            head = new_node
        lists.append(head)
    
    return lists


def allocate_hash_map(size: int, count: int) -> List[Dict[int, int]]:
    """Allocate hash map/dictionary structures."""
    maps = []
    for i in range(count):
        hash_map = {}
        for j in range(size):
            key = random.randint(0, size * 2)
            value = random.randint(0, 1000)
            hash_map[key] = value
        maps.append(hash_map)
    return maps


def allocate_sequential(allocator_func, size: int, count: int) -> Any:
    """Sequential allocation pattern."""
    return allocator_func(size, count)


def allocate_random(allocator_func, size: int, count: int) -> List[Any]:
    """Random allocation pattern."""
    # Allocate in random order with random delays
    allocated = []
    indices = list(range(count))
    random.shuffle(indices)
    
    for i in indices:
        # Allocate one item at a time
        item = allocator_func(size, 1)[0]
        allocated.append(item)
        
        # Random micro-delay to simulate real-world patterns
        if random.random() < 0.1:
            time.sleep(0.0001)
    
    return allocated


def allocate_fragmented(allocator_func, size: int, count: int) -> List[Any]:
    """Fragmented allocation pattern - allocate, deallocate some, allocate more."""
    allocated = []
    
    # First wave - allocate half
    first_half = count // 2
    first_batch = allocator_func(size, first_half)
    allocated.extend(first_batch)
    
    # Deallocate every other item to create fragmentation
    for i in range(len(allocated) - 1, -1, -2):
        del allocated[i]
    
    # Force garbage collection
    gc.collect()
    
    # Second wave - allocate remaining
    remaining = count - len(allocated)
    second_batch = allocator_func(size, remaining)
    allocated.extend(second_batch)
    
    return allocated


def run_memory_allocation_benchmark(config: Dict) -> Dict:
    """Run memory allocation benchmark."""
    allocation_sizes = config.get("allocation_sizes", [1024])
    patterns = config.get("allocation_patterns", ["sequential"])
    allocation_counts = config.get("allocation_counts", [100])
    data_structures = config.get("data_structures", ["array"])
    iterations = config.get("iterations", 3)
    
    # Structure allocators
    allocators = {
        "array": allocate_array,
        "linked_list": allocate_linked_list,
        "hash_map": allocate_hash_map
    }
    
    # Pattern functions
    pattern_funcs = {
        "sequential": allocate_sequential,
        "random": allocate_random,
        "fragmented": allocate_fragmented
    }
    
    results = {
        "start_time": time.time(),
        "test_cases": [],
        "summary": {
            "total_tests": 0,
            "successful_tests": 0,
            "failed_tests": 0,
            "avg_allocation_time": 0.0,
            "avg_deallocation_time": 0.0,
            "avg_memory_efficiency": 0.0
        }
    }
    
    all_allocation_times = []
    all_deallocation_times = []
    all_memory_efficiencies = []
    
    for size in allocation_sizes:
        for count in allocation_counts:
            for structure in data_structures:
                for pattern in patterns:
                    if structure not in allocators or pattern not in pattern_funcs:
                        continue
                    
                    print(f"Testing {structure} allocation: size={size}, count={count}, pattern={pattern}...", file=sys.stderr)
                    
                    test_case = {
                        "allocation_size": size,
                        "allocation_count": count,
                        "data_structure": structure,
                        "allocation_pattern": pattern,
                        "iterations": [],
                        "avg_allocation_time": 0.0,
                        "avg_deallocation_time": 0.0,
                        "avg_memory_efficiency": 0.0
                    }
                    
                    allocation_times = []
                    deallocation_times = []
                    memory_efficiencies = []
                    
                    for i in range(iterations):
                        print(f"  Iteration {i+1}/{iterations}...", file=sys.stderr)
                        
                        # Force garbage collection before test
                        gc.collect()
                        initial_memory = get_memory_usage()
                        
                        iteration_result = {
                            "iteration": i + 1,
                            "initial_memory": initial_memory,
                            "allocation": {},
                            "deallocation": {}
                        }
                        
                        results["summary"]["total_tests"] += 1
                        success = True
                        
                        try:
                            # Allocation phase
                            allocator = allocators[structure]
                            pattern_func = pattern_funcs[pattern]
                            
                            start_time = time.time()
                            
                            if pattern == "sequential":
                                allocated_data = pattern_func(allocator, size, count)
                            else:
                                allocated_data = pattern_func(allocator, size, count)
                            
                            allocation_time = (time.time() - start_time) * 1000  # ms
                            
                            # Measure memory after allocation
                            peak_memory = get_memory_usage()
                            memory_used = peak_memory - initial_memory
                            
                            # Calculate theoretical vs actual memory usage
                            if structure == "array":
                                # For NumPy, we can get the exact byte size
                                theoretical_size = sum(arr.nbytes for arr in allocated_data)
                            elif structure == "hash_map":
                                theoretical_size = size * count * 16  # Key-value pairs
                            else:  # linked_list
                                theoretical_size = size * count * 24  # Node overhead
                            
                            memory_efficiency = (theoretical_size / memory_used * 100) if memory_used > 0 else 0
                            
                            allocation_times.append(allocation_time)
                            all_allocation_times.append(allocation_time)
                            memory_efficiencies.append(memory_efficiency)
                            all_memory_efficiencies.append(memory_efficiency)
                            
                            iteration_result["allocation"] = {
                                "success": True,
                                "time_ms": allocation_time,
                                "memory_used": memory_used,
                                "peak_memory": peak_memory,
                                "memory_efficiency": memory_efficiency,
                                "items_allocated": count
                            }
                            
                            # Deallocation phase
                            start_time = time.time()
                            
                            # Clear references to trigger deallocation
                            if isinstance(allocated_data, list):
                                allocated_data.clear()
                            del allocated_data
                            
                            # Force garbage collection
                            gc.collect()
                            
                            deallocation_time = (time.time() - start_time) * 1000  # ms
                            final_memory = get_memory_usage()
                            
                            deallocation_times.append(deallocation_time)
                            all_deallocation_times.append(deallocation_time)
                            
                            iteration_result["deallocation"] = {
                                "success": True,
                                "time_ms": deallocation_time,
                                "final_memory": final_memory,
                                "memory_freed": peak_memory - final_memory
                            }
                            
                        except Exception as e:
                            success = False
                            iteration_result["allocation"]["success"] = False
                            iteration_result["allocation"]["error"] = str(e)
                        
                        if success:
                            results["summary"]["successful_tests"] += 1
                        else:
                            results["summary"]["failed_tests"] += 1
                        
                        test_case["iterations"].append(iteration_result)
                    
                    # Calculate averages for this test case
                    if allocation_times:
                        test_case["avg_allocation_time"] = sum(allocation_times) / len(allocation_times)
                    if deallocation_times:
                        test_case["avg_deallocation_time"] = sum(deallocation_times) / len(deallocation_times)
                    if memory_efficiencies:
                        test_case["avg_memory_efficiency"] = sum(memory_efficiencies) / len(memory_efficiencies)
                    
                    results["test_cases"].append(test_case)
    
    # Calculate overall summary
    if all_allocation_times:
        results["summary"]["avg_allocation_time"] = sum(all_allocation_times) / len(all_allocation_times)
    if all_deallocation_times:
        results["summary"]["avg_deallocation_time"] = sum(all_deallocation_times) / len(all_deallocation_times)
    if all_memory_efficiencies:
        results["summary"]["avg_memory_efficiency"] = sum(all_memory_efficiencies) / len(all_memory_efficiencies)
    
    results["end_time"] = time.time()
    results["total_execution_time"] = results["end_time"] - results["start_time"]
    
    return results


def main():
    """Main entry point for memory allocation test."""
    if len(sys.argv) < 2:
        print("Usage: python memory_allocation.py <config_file>", file=sys.stderr)
        sys.exit(1)
    
    config_file = sys.argv[1]
    
    try:
        with open(config_file, 'r') as f:
            config = json.load(f)
        
        parameters = config.get("parameters", {})
        results = run_memory_allocation_benchmark(parameters)
        
        print(json.dumps(results, indent=2))
        
    except FileNotFoundError:
        print(f"Error: Config file '{config_file}' not found", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in config file: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()